File: app/src/main/java/com/marceljurtz/lifecounter/Helper/Counter.java
package com.marceljurtz.lifecounter.Helper;

import android.os.Parcel;
import android.os.Parcelable;

public class Counter implements Parcelable {

    public Counter(String description, int ATK, int DEF) {
        this.description = description;
        this.atk = ATK;
        this.def = DEF;
    }

    public Counter(String identifier, String description, int ATK, int DEF) {
        this.identifier = identifier;
        this.description = description;
        this.atk = ATK;
        this.def = DEF;
    }

    private Counter(Parcel source) {
        identifier = source.readString();
        description = source.readString();
        atk = source.readInt();
        def = source.readInt();
    }

    // ID to link counterToAdd to view
    // template: playerID_increment
    private String identifier;

    public String getIdentifier() {
        return identifier;
    }

    public void setIdentifier(String identifier) {
        this.identifier = identifier;
    }

    private String description;

    public String getDescription() {
        return this.description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    private int atk;

    public int getATK() {
        return this.atk;
    }

    public void setATK(int atk) {
        this.atk = atk;
    }

    private int def;

    public int getDEF() {
        return this.def;
    }

    public void setDEF(int def) {
        this.def = def;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(identifier);
        dest.writeString(description);
        dest.writeInt(atk);
        dest.writeInt(def);
    }

    public static final Parcelable.Creator<Counter> CREATOR
            = new Parcelable.Creator<Counter>() {
        @Override
        public Counter createFromParcel(Parcel in) {
            return new Counter(in);
        }

        @Override
        public Counter[] newArray(int size) {
            return new Counter[size];
        }
    };
}

File: app/src/main/java/com/marceljurtz/lifecounter/models/Counter.kt
package com.marceljurtz.lifecounter.models

import android.os.Parcel
import android.os.Parcelable

import com.marceljurtz.lifecounter.enums.CounterType

class Counter : Parcelable {

    // ID to link counterToAdd to view
    // template: playerID_increment
    var identifier: String? = null

    var description: String? = null

    var atk: Int = 0

    var def: Int = 0

    var counterType: CounterType? = null

    constructor(description: String, ATK: Int, DEF: Int, counterType: CounterType) {
        this.description = description
        this.atk = ATK
        this.def = DEF
        this.counterType = counterType
    }

    constructor(identifier: String, description: String, ATK: Int, DEF: Int, counterType: CounterType) {
        this.identifier = identifier
        this.description = description
        this.atk = ATK
        this.def = DEF
        this.counterType = counterType
    }

    private constructor(source: Parcel) {
        identifier = source.readString()
        description = source.readString()
        atk = source.readInt()
        def = source.readInt()
        counterType = CounterType.valueOf(source.readString())
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(identifier)
        dest.writeString(description)
        dest.writeInt(atk)
        dest.writeInt(def)
        dest.writeString(this.counterType!!.name)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<Counter> = object : Parcelable.Creator<Counter> {
            override fun createFromParcel(`in`: Parcel): Counter {
                return Counter(`in`)
            }

            override fun newArray(size: Int): Array<Counter?> {
                return arrayOfNulls(size)
            }
        }
    }
}
